docker/lr-pb/shard_bam.py

```diff
@@ -0,0 +1,137 @@
+import argparse
+import gzip
+from collections import OrderedDict
+from math import ceil
+import pysam
+from construct import *
+
+from multiprocessing.pool import ThreadPool
+from functools import partial
+
+
+def write_shard(bam, sharding_offsets, zmw_counts_exp, prefix, index):
+    """
+    Write subsection of PacBio bam to a shard.
+    """
+
+    bf = pysam.Samfile(bam, 'rb', check_sq=False)
+
+    # Advance to the specified virtual file offset.
+    bf.seek(sharding_offsets[index])
+
+    num_reads = 0
+    zmw_counts_act = {}
+    with pysam.Samfile(f'{prefix}{index}.bam', 'wb', header=bf.header) as out:
+        # Write until we've advanced to (but haven't written) the read that begins the next shard.
+        while True:
+            read = bf.__next__()
+            out.write(read)
+
+            # Count the ZMW numbers seen.
+            zmw = read.get_tag("zm")
+            zmw_counts_act[zmw] = zmw_counts_act.get(zmw, 0) + 1
+            num_reads += 1
+
+            if bf.tell() >= sharding_offsets[index+1]:
+                break
+
+    # Verify that the count for ZMWs written to this shard match the ZMW count determined
+    # from the original read of the index file.  If these assertions fail, it may indicate
+    # that reads from the same ZMW have been erroneously sharded to separate files.
+    for zmw in zmw_counts_act:
+        assert zmw_counts_act[zmw] == zmw_counts_exp[zmw]
```

Comment:
I don't think (but not strongly) this check should be here, as long as the test below is not buggy, this should never happen and this assertion check has performance penalties.

Reply:
Good point.  I still do want a check here (in case someone supplies the wrong .pbi file for a .bam file).  I've changed this to just raising an exception.